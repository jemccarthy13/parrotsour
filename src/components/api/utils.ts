export async function validAccessCode(accessCode: string) {
  const resp = await fetch(
    `${process.env.PUBLIC_URL}/database/codes/${accessCode}.php`,
    {
      method: "POST",
    }
  )

  return resp.status === 202
}
