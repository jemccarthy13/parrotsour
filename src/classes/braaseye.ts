import { BRAA } from "classes/braa"
import { Point } from "classes/point"

export class Braaseye {
  bull: BRAA
  braa: BRAA
  private drawLoc: Point

  constructor(toPoint: Point, fromPoint: Point, bullseye: Point) {
    this.bull = bullseye.getBR(toPoint)
    this.braa = fromPoint.getBR(toPoint)
    this.drawLoc = toPoint
  }

  /**
   * Draw BRAASEYE, bearing and range to point from two reference points
   * (typically aircraft to target and bullseye to target)
   *
   * @param showMeasurements true iff braaseye should be drawn on drawing context
   * @param braaFirst true iff BRAA is displayed first, false to draw bullseye first
   * @param offsetX X position to draw at
   * @param offsetY Y position to draw at
   */
  draw(
    showMeasurements: boolean,
    braaFirst: boolean,
    offsetX = 0,
    offsetY = 0
  ): void {
    if (braaFirst) {
      this.braa.draw(
        this.drawLoc.x + 25 + offsetX,
        this.drawLoc.y + offsetY,
        "blue",
        showMeasurements
      )
      this.bull.draw(
        this.drawLoc.x + 25 + offsetX,
        this.drawLoc.y + 11 + offsetY,
        "black",
        showMeasurements
      )
    } else {
      this.bull.draw(
        this.drawLoc.x + 25 + offsetX,
        this.drawLoc.y + offsetY,
        "black",
        showMeasurements
      )
      this.braa.draw(
        this.drawLoc.x + 25 + offsetX,
        this.drawLoc.y + 11 + offsetY,
        "blue",
        showMeasurements
      )
    }
  }
}
