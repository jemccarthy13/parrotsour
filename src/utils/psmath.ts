import { FORMAT } from "../classes/supportedformats"

export const PIXELS_TO_NM = 4

/**
 * Converts a math angle to Radians (180 is EW line)
 * @param angleDeg - the cartesian angle to convert to radians
 */
export function toRadians(angleDeg: number): number {
  return angleDeg * (Math.PI / 180)
}

/**
 * Converts radians to cartesian degrees
 * @param rads - radians to convert to degrees
 */
export function toDegrees(rads: number): number {
  return rads * (180 / Math.PI)
}

/**
 * Convert an aeronautical heading to Cartesian radians,
 * and return useful information about that heading.
 *
 * @param heading heading in degrees
 * @returns An object containing:
 *          - radians (heading -> radians)
 *          - offset (90 deg perpendicular to heading, useful for
 *              spacing DataTrails)
 *          - headAngle (150 deg delta from heading, useful for
 *              drawing Arrow DataTrail)
 */
export function headingToRadians(heading: number): {
  radians: number
  headAngle: number
  offset: number
} {
  let deg: number = 360 - (heading - 90)
  if (heading < 90) deg = 90 - heading

  let arrowHead = deg - 150
  if (arrowHead < 0) arrowHead = 360 + arrowHead

  let offsetVector = deg - 90
  if (offsetVector < 0) offsetVector = 360 + offsetVector

  return {
    radians: toRadians(deg),
    headAngle: toRadians(arrowHead),
    offset: toRadians(offsetVector),
  }
}

/**
 * Returns a random number from min to max (inclusive both)
 * @param min minimum value
 * @param max maximum value
 */
export function randomNumber(min: number, max: number): number {
  return Math.floor(Math.random() * (max - min + 1) + min)
}

/**
 * Return a random heading (0-360 for ALSA, HOT/FLANK blue for !ALSA)
 * @param format
 */
export function randomHeading(format: FORMAT, blueHeading = -1): number {
  const bound = format === FORMAT.IPE ? 45 : 360
  const offset = randomNumber(-bound, bound)

  let blueOpp = blueHeading - 180
  if (blueOpp < 0) blueOpp = 360 - blueOpp

  let heading: number = blueOpp + offset
  heading = (360 + heading) % 360

  return heading
}

/**
 * Get the shortest delta between two angles (i.e. do I turn left or
 * right to get to the 'desired' angle)
 */
export function getDegDeltaBetween(
  heading: number,
  desHeading: number,
  forced: string | undefined
): number {
  const LH = (heading - desHeading + 360) % 360
  const RH = (desHeading - heading + 360) % 360

  let deltaA = RH
  if (forced === "LEFT") {
    deltaA = -LH
  } else if (LH < RH) {
    deltaA = -LH
  }
  return deltaA
}
