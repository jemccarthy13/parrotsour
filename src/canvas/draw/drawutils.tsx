/**
 * This file contains common low-level drawing utilities used
 * to build groups and pictures.
 */

// Interfaces
import { Point } from "../../classes/point"

/**
 * 'Clamp' the location to the confines of the drawing context
 * @param ctx The context to constrict it to
 * @param pos the current position
 */
export function clampInContext(
  ctx: CanvasRenderingContext2D,
  x: number | Point,
  y?: number
): Point {
  let retPoint = Point.DEFAULT
  if (x instanceof Point) {
    retPoint = new Point(
      Math.min(Math.max(x.x, 0), ctx.canvas.width),
      Math.min(Math.max(x.y, 0), ctx.canvas.height)
    )
  } else {
    if (y === undefined) y = 0
    retPoint = new Point(
      Math.min(Math.max(x, 0), ctx.canvas.width),
      Math.min(Math.max(y, 0), ctx.canvas.height)
    )
  }
  return retPoint
}
