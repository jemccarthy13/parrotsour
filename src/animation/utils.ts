import { PaintBrush } from "../canvas/draw/paintbrush"
import { SensorType } from "../classes/aircraft/datatrail/sensortype"
import { AircraftGroup } from "../classes/groups/group"

/**
 * Internal function to see if a group is near the canvas boundary.
 *
 * Used to change group logic when approaching the edge of the canvas
 * @param group The group to check against boundaries
 * @param dataStyle The current DataTrail type
 */
export function isNearBounds(
  group: AircraftGroup,
  dataStyle: SensorType
): boolean {
  const buffer = 40
  const sX = group.getCenterOfMass(dataStyle).x
  const sY = group.getCenterOfMass(dataStyle).y
  const ctx = PaintBrush.getContext()

  return (
    sX < buffer ||
    sX > ctx.canvas.width - buffer ||
    sY < buffer ||
    sY > ctx.canvas.height - buffer
  )
}
