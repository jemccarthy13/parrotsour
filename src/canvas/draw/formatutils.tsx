/**
 * This file contains utilities for group and answer formatting
 */
import { AircraftGroup } from "../../classes/groups/group"
import { toRadians } from "../../utils/psmath"
import { ArrowDataTrail } from "../../classes/aircraft/datatrail/arrowdatatrail"

/**
 * Convert an altitude to a 3-digit flight level
 * @param alt int altitude to change to padded string
 */
export function formatAlt(alt: number): string {
  return (alt * 10).toString().substring(0, 3).padStart(3, "0")
}

/**
 * Check if picture is opening or closing
 * @param fg First group of picture
 * @param sg Second group of picture
 */
export function getOpenCloseAzimuth(
  fg: AircraftGroup,
  sg: AircraftGroup
): string {
  let retVal = ""

  const fgStartPos = fg.getStartPos()
  const sgStartPos = sg.getStartPos()

  const angle1to2 = fgStartPos.getBR(sgStartPos).bearingNum

  const fgHeading = fg.getHeading()
  const sgHeading = sg.getHeading()

  const speed1to2 =
    ArrowDataTrail.LEN_TRAIL * Math.cos(toRadians(angle1to2 - fgHeading))

  const speed2to1 =
    ArrowDataTrail.LEN_TRAIL * Math.cos(toRadians(angle1to2 - sgHeading))

  const acceptableRate = Math.cos(toRadians(5))

  let closing = speed1to2 > 0 && speed2to1 < 0
  let opening = speed1to2 < 0 && speed2to1 > 0

  closing =
    closing ||
    (Math.abs(speed1to2) < acceptableRate && speed2to1 < 0) ||
    (Math.abs(speed2to1) < acceptableRate && speed1to2 > 0)
  opening =
    opening ||
    (Math.abs(speed1to2) < acceptableRate && speed2to1 > 0) ||
    (Math.abs(speed2to1) < acceptableRate && speed1to2 < 0)

  const parallel =
    Math.abs(speed1to2) < acceptableRate && Math.abs(speed2to1) < acceptableRate

  if (!parallel) {
    if (opening) {
      retVal = " OPENING "
    }
    if (closing) {
      retVal = " CLOSING "
    }
  }

  return retVal
}
