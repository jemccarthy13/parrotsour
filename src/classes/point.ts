import { BlueInThe } from "../canvas/canvastypes"
import { BRAA } from "./braa"
import { PIXELS_TO_NM, toDegrees } from "../utils/psmath"

/**
 * An x,y Cartesian point that can calculate bearing and
 * range to another point.
 */
export class Point {
  x: number
  y: number

  static DEFAULT = new Point(0, 0)

  /**
   * Construct a new x,y point
   */
  constructor(x: number, y: number) {
    this.x = x
    this.y = y
  }

  /**
   * Get a bearing and range between this point and another point
   *
   * @param ptTo the "other" or "to" point
   * @returns the bearing and range between this (ptFrom) and ptTo
   */
  getBR(toPoint: Point): BRAA {
    const deltaX = this.x - toPoint.x
    const deltaY = this.y - toPoint.y

    // distance formula for range
    const rng = Math.floor(
      Math.sqrt(deltaX * deltaX + deltaY * deltaY) / PIXELS_TO_NM
    )

    // convert cartesian angle to heading
    let brg = Math.round(
      270 + toDegrees(Math.atan2(this.y - toPoint.y, this.x - toPoint.x))
    )
    if (brg > 360) {
      brg = brg - 360
    }

    return new BRAA(brg, rng)
  }

  /**
   * Get distance between points (straight along axis, i.e. not slant range)
   */
  public straightDistNM(toPoint: Point, orientation: BlueInThe): number {
    return Math.floor(
      Math.abs(
        orientation === BlueInThe.NORTH || orientation === BlueInThe.SOUTH
          ? this.y - toPoint.y
          : this.x - toPoint.x
      ) / PIXELS_TO_NM
    )
  }
}
