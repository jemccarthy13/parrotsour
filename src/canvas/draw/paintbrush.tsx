import { Braaseye } from "../../classes/braaseye"
import { AircraftGroup } from "../../classes/groups/group"
import { Point } from "../../classes/point"
import { PIXELS_TO_NM, randomNumber } from "../../utils/psmath"
import { PictureCanvasProps, PictureCanvasState } from "../canvastypes"
import { clampInContext } from "./drawutils"
import { formatAlt } from "./formatutils"

export class PaintBrush {
  private static ctx: CanvasRenderingContext2D

  constructor(ctx: CanvasRenderingContext2D) {
    PaintBrush.ctx = ctx
  }

  public static use(ctx: CanvasRenderingContext2D | undefined | null): void {
    if (ctx) PaintBrush.ctx = ctx
  }

  public static getContext(): CanvasRenderingContext2D {
    return PaintBrush.ctx
  }

  public static clearCanvas(): void {
    const ctx = PaintBrush.ctx
    ctx.clearRect(0, 0, ctx.canvas.width, ctx.canvas.height)
    ctx.fillStyle = "white"
    ctx.fillRect(0, 0, ctx.canvas.width, ctx.canvas.height)
  }

  /**
   * Draw a line on the drawing context, given the properties
   * @param startX start x of the line
   * @param startY start y of the line
   * @param endX end x of the line
   * @param endY end y of the line
   * @param color (optional) color of the line, defaults to black
   */
  public static drawLine(
    startX: number,
    startY: number,
    endX: number,
    endY: number,
    color = "black",
    forcedCtx?: CanvasRenderingContext2D
  ): void {
    let context = PaintBrush.ctx
    if (forcedCtx) context = forcedCtx
    context.lineWidth = 1
    context.strokeStyle = color
    context.beginPath()
    context.moveTo(startX, startY)
    context.lineTo(endX, endY)
    context.stroke()
    context.stroke()
  }

  /**
   * Draw text to the drawing context at the given position
   * @param text Text to draw
   * @param x X Position to draw at
   * @param y Y Position to draw at
   * @param size (optional) Size of the text to draw, defaults 12
   * @param color (optional) Color of the text, defaults to black
   */
  public static drawText(
    text: string,
    x: number,
    y: number,
    size = 12,
    color = "black"
  ): void {
    const ctx = PaintBrush.ctx
    ctx.lineWidth = 1
    ctx.fillStyle = color
    ctx.font = size + "px Arial"
    const pos = clampInContext(ctx, x, y)

    ctx.fillText(text, pos.x, pos.y)
    ctx.fillText(text, pos.x, pos.y)
  }

  /**
   * Draw a measurement (distance with a length number)
   * @param {Point} start position to start for line
   * @param {Point} end position to end for line
   * @param {number} distance how long the line should be
   * @param {boolean} showMeasurements Display the measurements
   */
  public static drawMeasurement(
    start: Point,
    end: Point,
    distance: number,
    showMeasurements: boolean
  ): void {
    if (showMeasurements) {
      PaintBrush.drawLine(start.x, start.y, end.x, end.y)
      PaintBrush.drawText(
        Math.floor(distance).toString(),
        (start.x + end.x) / 2,
        (start.y + end.y) / 2 - 3
      )
    }
  }

  /**
   * Draw altitudes next to a group with optional offset
   *
   * @param grpPos group's current position
   * @param alts altitudes to draw
   * @param offX (optional) x-axis offset from group
   * @param offY (optional) y-axis offset from group
   */
  public static drawAltitudes(
    grpPos: Point,
    alts: number[],
    offX?: number,
    offY?: number
  ): void {
    const offsetX = offX || 0
    const offsetY = offY || 0
    const formattedAlts: string[] = alts.map((a: number) => {
      return formatAlt(a)
    })
    PaintBrush.drawText(
      formattedAlts.join(","),
      grpPos.x + 25 + offsetX,
      grpPos.y - 11 + offsetY,
      11,
      "#ff8c00"
    )
  }

  public static drawBullseye(bull?: Point, color?: string): Point {
    const context = PaintBrush.ctx
    color = color || "black"
    context.lineWidth = 1
    context.fillStyle = color
    context.strokeStyle = color

    const centerPointX = bull
      ? bull.x
      : randomNumber(context.canvas.width * 0.33, context.canvas.width * 0.66)
    const centerPointY = bull
      ? bull.y
      : randomNumber(context.canvas.height * 0.33, context.canvas.height * 0.66)

    context.beginPath()
    context.arc(
      centerPointX,
      centerPointY,
      PIXELS_TO_NM / 2,
      0,
      2 * Math.PI,
      true
    )
    context.stroke()
    context.fill()

    context.moveTo(centerPointX, centerPointY + PIXELS_TO_NM * 2)
    context.lineTo(centerPointX, centerPointY - PIXELS_TO_NM * 2)
    context.stroke()

    context.moveTo(centerPointX + PIXELS_TO_NM * 2, centerPointY)
    context.lineTo(centerPointX - PIXELS_TO_NM * 2, centerPointY)
    context.stroke()

    return new Point(centerPointX, centerPointY)
  }

  public static drawFullInfo(
    state: PictureCanvasState,
    props: PictureCanvasProps,
    groups: AircraftGroup[]
  ): void {
    for (let y = 0; y < groups.length; y++) {
      const grpPos = groups[y].getCenterOfMass(props.dataStyle)
      if (props.showMeasurements) {
        new Braaseye(
          grpPos,
          state.blueAir.getCenterOfMass(props.dataStyle),
          state.bullseye
        ).draw(true, props.braaFirst)
      }
      PaintBrush.drawAltitudes(grpPos, groups[y].getAltitudes())
    }
  }
}
