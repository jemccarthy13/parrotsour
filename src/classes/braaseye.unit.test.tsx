import { PaintBrush } from "../canvas/draw/paintbrush"
import { PIXELS_TO_NM } from "../utils/psmath"
import { Braaseye } from "./braaseye"
import { Point } from "./point"

// eslint-disable-next-line @typescript-eslint/no-var-requires
const canvasSerializer = require("jest-canvas-snapshot-serializer")

expect.addSnapshotSerializer(canvasSerializer)

describe("Braaseye", () => {
  const canvas = document.createElement("canvas")
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const ctx = canvas.getContext("2d")!
  canvas.height = 50
  canvas.width = 100
  PaintBrush.use(ctx)

  afterEach(() => {
    canvas.height = 50
    canvas.width = 100
    PaintBrush.getContext().fillStyle = "white"
    PaintBrush.getContext().fillRect(0, 0, canvas.width, canvas.height)
  })

  const bluePos = new Point(10, 10)
  const bullseye = new Point(5, 10)
  const toPoint = new Point(20, 10)
  it("constructs_correctly", () => {
    const braaseye = new Braaseye(toPoint, bluePos, bullseye)
    expect(braaseye.braa.bearing).toEqual("090")
    expect(braaseye.braa.bearingNum).toEqual(90)
    expect(braaseye.braa.range).toEqual(Math.floor(10 / PIXELS_TO_NM))
    expect(braaseye.bull.bearing).toEqual("090")
    expect(braaseye.bull.bearingNum).toEqual(90)
    expect(braaseye.bull.range).toEqual(Math.floor(15 / PIXELS_TO_NM))
  })

  it("draws_correctly_noOffset", () => {
    const braaseye = new Braaseye(toPoint, bluePos, bullseye)
    braaseye.draw(true, true)
    expect(canvas).toMatchSnapshot()
  })

  it("draws_correctly_withOffset", () => {
    const braaseye = new Braaseye(toPoint, bluePos, bullseye)
    braaseye.draw(true, true, -10, 10)
    expect(canvas).toMatchSnapshot()
  })

  it("draws_correctly_braaFirst", () => {
    const braaseye = new Braaseye(toPoint, bluePos, bullseye)
    braaseye.draw(true, false, -10, 10)
    expect(canvas).toMatchSnapshot()
  })
})
