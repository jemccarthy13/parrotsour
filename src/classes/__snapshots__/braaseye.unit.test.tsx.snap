// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Braaseye draws_correctly_braaFirst 1`] = `
<canvas
  data-snapshot-image="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAAAyCAYAAACqNX6+AAAABmJLR0QA/wD/AP+gvaeTAAAD20lEQVR4nO2ZsUtbbRSHf++X3ERj0i6KRChkSKX+F4Iuit0U2q2Y0YJDKDgUBF3V7KZDseBsBwfnOIlgi7vEIBIpcisiLm2fDjdqbL8E+/XTnOF9QHhz+Rly7nPve3JPHIA8Zvin0x/AcxsvxBheiDG8EGN4IcbwQozhhRjDCzGGF2IML8QYXogxTAkpl8tKJpNyzqmnp0fLy8uSpM3NTaXTaTnnNDAwoMPDw7Z5Sdre3lYQBNrY2FAmk5FzTul0WktLSx2p7c5giGw2y+TkJPV6nZGREfr7+wHIZDIUi0XCMGRhYYGhoaG2eYBCoUA+n6e7u5vp6WmOj4+Zn58nHo93pLa7YkpILBajWq0CsLu7SxAEAEiiXq8DcHJyQiwWa5sHyOfzFAqF69e1Wo1CocCjR48epJb/iikhkri8vATg7OwM5xwAqVSKxcVFwjBkamqKqxu7VR4gHo9TqVQAODo6wjmHJGZnZx+ypD/GlBDnHGdnZwCEYXh9gtfX1+nq6sI5x+jo6PXxVvlKpfKvW9PMzMytu8gipoQEQcD+/j4Ae3t7JJPJ3zJbW1t0dXW1zV/1j1+p1+sYa5u/YerT5XI5JiYmOD09ZXx8nGw2C0Qnfm5ujoODA8bGxhgeHm6bb+4fiUSClZUVLi4uKBaL9PX1daa4O2JKSKlUIhaLIYlEIsHq6ur18SAIcM4xODjI+fl523xz/1hbW6O7uxtJ9Pb2srOz05ni7ogD/5u6JUw9GHq8EHOYElIuS8mk5JzU0yNdTUI2N6V0Ojo+MCA1Jict85K0vS0FgfTunZRKRZlUSvrw4eHr+iM63cSayWZhchLqdRgZgatJSCYDxSKEISwsQGNy0jIPUChAPg+JBLx9G/3vixfQ+MZsFlNCYjFoTELY3YWrZzgpOukAJydRrl0eIhlNkxMAvnyBpod5k5jasr5/l/r7o/XTp9K3b9E6lYq2p69fpdevo1y7vCRVq9KrV7ffv1yWnj27zwr+Bzp9RTTjHDQmIYThzdW8vh5tNc7B6OjN8Vb5SgV+nZx8/gy9vVCr3X8df4OpOyQel2q1aF2tSolEtH75Urq8lH78kN68iRp5u/z791Iud/O+nz5Jz59LHz9KT57cfx1/RaeviGZyOZiYgNNTGB+PmjZEvWFuDg4OYGwMGpOTlvnm/rG4CI8fQ2PkZR5TQkqlqFFL0bejxiSEUimS4hwMDkJjctIyH49H29bVWrr9Zxk/OjGGqR7i8ULM4YUYwwsxhhdiDC/EGF6IMbwQY3ghxvBCjOGFGMMLMYYXYgwvxBheiDG8EGN4IcbwQozxEz9Ug5J6UgbtAAAAAElFTkSuQmCC"
  height="50"
  width="100"
/>
`;

exports[`Braaseye draws_correctly_noOffset 1`] = `
<canvas
  data-snapshot-image="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAAAyCAYAAACqNX6+AAAABmJLR0QA/wD/AP+gvaeTAAAD3UlEQVR4nO2Yv2taXxjGn/P1Xn+bLgZRKDhYaf6LgC6K3Qy0W8kdU8ggBYdCQNck7rFDSSFzOjhk1ikIackuRkJQitxKCC5tnw7nJtF8o0uJnlvOB4TD4fHi6+d6Xu8rSBIaZfhv2R9AM40rhdTrgM8HCAGEQsDentxvNIBwWO4nEsDFxfw8ALRagGkCHz8CwaDMBIPA58+LrwsAQBcSj5PFItnvk5kMGYvJ/UiELJVI2yYrFXJtbX6eJC2LTKVIr5f88EG+9/Vr0u9ffF0k6UohHg/Z7cp1u02aplwD8ksnycFA5ublSSnDsqav//07KcTTff55uPLI+vULiMXk+sUL4OdPuQ4G5fH04wfw7p3MzcsDQLcLvH07ff16HXj58ikrmMNy7oO/QwhyNJJr276/m4+O5FEjBJnN3u/PyjebpGFMX/vbNzIaJXu9p6/jMVz5CzEMoNeT624X8Hrl+s0bYDwGfv8G3r+XjXxe/tMnIJm8v+7Xr8CrV8CXL8Dz509fx6Ms5z74O5JJslAgh0Myn5dNm5S9oVwmOx0ylyPX1+fnJ/tHtUo+e0aeny++nklcKaRWk40akP+ODg7u901THknpNHl9PT9vGPLYul0D069lIEj9pK4Sruwh/zJaiGK4Uki9XofP54MQAqFQCHvOLKTRaCAcDkMIgUQigQtndjIrDwCtVgumaeL4+BiRSARCCITDYezu7i6lNlc29Xg8zmKxyH6/z0wmw5gzC4lEIiyVSrRtm5VKhWvO7GRWniQty2IqlWIgEODm5iavrq64s7ND4+EDyoJwpRCPx8OuMwtpt9s0nVkIAPad2clgMKDHmZ3MypNkKpWiNTE76fV6tCyLKysrC6nlIa4UAoDj8ZgkORqNKJxH72AwyGq1Stu2ubGxwdsDYFaeJA3DYNP573t5eUkhBAFwe3t7kSXd4UohQgiOnFmIbdt3X/DR0RH9fj+FEMxms3f7s/LNZvPRo2lra2vqV7RIXCnENE2eO4/UZ2dn9Pl8/8ucnJzQ78zQZ+Vv+8dD+v0+l9VeXSkkmUyyUChwOBwyn88z7sxCTNNkuVxmp9NhLpfjujM7mZWf7B9er5f7+/u8ublhqVTi6urqUmpzpZBarUaPx0MA9Hq9PHBmIbVajaZpUgjBdDrNa2d2Mis/2T8ODw8ZCAQIgNFolKenp0upTY9OFMOVD4b/MlqIYmghiqGFKIYWohhaiGJoIYqhhSiGFqIYWohiaCGKoYUohhaiGFqIYmghiqGFKIYWohhaiGJoIYqhhSiGFqIYWohiaCGKoYUohhaiGFqIYmghiqGFKMYfl86DknDU+1sAAAAASUVORK5CYII="
  height="50"
  width="100"
/>
`;

exports[`Braaseye draws_correctly_withOffset 1`] = `
<canvas
  data-snapshot-image="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAAAyCAYAAACqNX6+AAAABmJLR0QA/wD/AP+gvaeTAAAD40lEQVR4nO2Zv0sbbxzH3883d5ffdolIAoUMaaj/haBLJN0itFvxRgsOoZChICSrmt10KBac7eDgnEwSsMVdYhBJKOEaRLK0fXd4zpq03wRtq/kMzwuEh4e3x33yurtP7hNFkjCI4b9pn4BhFCNEGEaIMIwQYRghwjBChGGECMMIEYYRIgwjRBhGiDBECanVgGAQUAqIRoGtLb1/cADEYno/lQLOzibnAaDRAGwbePsWiER0JhIB3r9/+LruBAWRTJKFAtnpkIuL5Nyc3o/HyWKR9DyyXCbn5yfnSdJ1yUyGdBzyzRv9v8+fk6HQw9d1F0QJCQTIVkuvm03StvUa0B86SXa7OjcpT2oZrjt6/M+fSaXu7/z/BaIeWd++AXNzev3kCfD1q15HIvrx9OUL8OqVzk3KA0CrBbx8OXr8Wg14+vQ+K/gHTPuKGEYpst/Xa8+7uZr39vSjRilyaelmf1y+Xicta/TYnz6RiQTZbt9/HX+DqDvEsoB2W69bLcBx9PrFC2AwAL5/B16/1o18Uv7dOyCdvjnux4/As2fAhw/A48f3X8dfMe0rYph0msznyV6PXF7WTZvUvaFUIk9PyVyOXFiYnB/uH5UK+egReXLy8PX8CaKEVKu6UQP629HOzs2+betHUjZLXl5OzluWfmxdr4HRP8ko0vymLglRPcRghIhDlJBarYZgMAilFKLRKLb8WcjBwQFisRiUUkilUjjzZyfj8gDQaDRg2zb29/cRj8ehlEIsFsPm5uZUars1025iwySTSRYKBXY6HS4uLnLOn4XE43EWi0V6nsdyucx5f3YyLk+Srusyk8kwHA5zdXWVFxcX3NjYoPXrC4owRAkJBAJs+bOQZrNJ25+FAGDHn510u10G/NnJuDxJZjIZukOzk3a7Tdd1OTMz8yC1/CmihADgYDAgSfb7fSr/1TsSibBSqdDzPK6srPD6xh6XJ0nLslj3v/uen59TKUUAXF9ff8iS7owoIUop9v1ZiOd5Pz/gvb09hkIhKqW4tLT0c39cvl6v/++jaW1tbeQukogoIbZt88R/pT4+PmYwGPwtc3h4yJA/Qx+Xv+4fv9LpdCisbf6GqLNLp9PM5/Ps9XpcXl5m0p+F2LbNUqnE09NT5nI5Lvizk3H54f7hOA63t7d5dXXFYrHI2dnZ6RR3S0QJqVarDAQCBEDHcbjjz0Kq1Spt26ZSitlslpf+7GRcfrh/7O7uMhwOEwATiQSPjo6mU9wtMaMTYYh6MTQYIeIwQoRhhAjDCBGGESIMI0QYRogwjBBhGCHCMEKEYYQIwwgRhhEiDCNEGEaIMIwQYRghwvgBJkuDkiH63kQAAAAASUVORK5CYII="
  height="50"
  width="100"
/>
`;
